More decoupling

davidfowl opened the pull request:
- Remove Libuv and Hosting dependency from Kestrel.Core
- Kestrel.Core depends on Hosting.Abstractions

cesarbs commented:
Cool, except for `KestrelServer : IServerAddressesFeature`.

davidfowl commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Microsoft.AspNetCore.Server.Kestrel.Core.csproj:
```diff
@@ -13,9 +13,11 @@
-    <PackageReference Include="Microsoft.AspNetCore.Hosting" Version="$(AspNetCoreVersion)" />
+    <PackageReference Include="Microsoft.AspNetCore.Hosting.Abstractions" Version="$(AspNetCoreVersion)" />
+    <PackageReference Include="Microsoft.AspNetCore.WebUtilities" Version="$(AspNetCoreVersion)" />
+    <PackageReference Include="Microsoft.Net.Http.Headers" Version="$(AspNetCoreVersion)" />
```
I blame @JunTaoLuo for this

Tratcher commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/ServerAddressesFeature.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
+{
+    internal class ServerAddressesFeature : IServerAddressesFeature
```
heads up @JunTaoLuo

davidfowl replied:
😄

halter73 replied:
Could this class just be moved to Hosting.Abstractions?

davidfowl replied:
Sure after this change.

Tratcher replied:
I'd rather you clone it than move it to Abstractions

JunTaoLuo replied:
@davidfowl you'll need to update this after https://github.com/aspnet/Hosting/pull/992 😄

davidfowl commented on src/Microsoft.AspNetCore.Server.Kestrel.Https/Microsoft.AspNetCore.Server.Kestrel.Https.csproj:
```diff
@@ -16,6 +16,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.Http" Version="$(AspNetCoreVersion)" />
```
This could be Hosting.Abstractions

halter73 replied:
What type(s) was this reference added for? Was it being brought in transitively through Core's reference to Hosting before?

davidfowl replied:
https://github.com/aspnet/HttpAbstractions/blob/a0bb843dd581c1521ad8bcbe21923959158cc1f2/src/Microsoft.AspNetCore.Http/Features/TlsConnectionFeature.cs

halter73 replied:
So if you copied TlsConnectionFeature.cs like you did for ServerAddressesFeature.cs, you could get rid of this package reference decouple even more.

davidfowl replied:
Sure, I'm fine with that.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel/Microsoft.AspNetCore.Server.Kestrel.csproj:
```diff
@@ -12,6 +12,10 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.Hosting" Version="$(AspNetCoreVersion)" />
```
This is needed, right? The only class in this project, WebHostBuilderKestrelExtension, only seems to reference IWebHostBuilder, IServer and some config stuff. That should come from Hosting.Abstractions, Hosting.Server.Abstractions and Options respectively.

All of this should come in transitively. If that isn't the case, why is the Libuv project building which also has IWebHostBuilder extensions?

Tratcher replied:
Kestrel pulls it in so the app doesn't have to.

davidfowl replied:
This is to not break upstream components that might have been relying on Kestrel depending on Hosting. I don't want to change that as part of this. The goal here was to clean up some of the dependencies after the initial change.

halter73 replied:
What if the app wants to use a non-libuv transport? Maybe we should keep the reference in Core then.

cesarbs commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/ServerAddressesFeature.cs:
```diff
@@ -0,0 +1,12 @@
+﻿using System;
```
Copyright header.

davidfowl commented:
Spoke to @halter73 offline. I've added a private `TlsConnectionFeature` in the https connection adapter and added copyrights. This is the first stage to further dependency decoupling.
